src/usb.c

```diff
@@ -1025,50 +1025,44 @@ stlink_t *stlink_open_usb(enum ugly_loglevel verbose, bool reset, char serial[ST
             }
         }
 
-        if ((desc.idProduct == STLINK_USB_PID_STLINK_32L)
-                || (desc.idProduct == STLINK_USB_PID_STLINK_NUCLEO)
-                || (desc.idProduct == STLINK_USB_PID_STLINK_32L_AUDIO)
-                || (desc.idProduct == STLINK_USB_PID_STLINK_V3_USBLOADER)
-                || (desc.idProduct == STLINK_USB_PID_STLINK_V3E_PID)
-                || (desc.idProduct == STLINK_USB_PID_STLINK_V3S_PID)
-                || (desc.idProduct == STLINK_USB_PID_STLINK_V3_2VCP_PID)) {
-            struct libusb_device_handle *handle;
-
-            ret = libusb_open(list[cnt], &handle);
-            if (ret)
-                continue;
+		ret = libusb_open(list[cnt], &handle);
+
+		/* could not open device, continue */
+		if (ret)
+			continue;
+
 
 		sl->serial_size = libusb_get_string_descriptor_ascii(handle, desc.iSerialNumber,
 				(unsigned char *)sl->serial, sizeof(sl->serial));
 
-            if ((desc.idProduct == STLINK_USB_PID_STLINK_32L)
-                    || (desc.idProduct == STLINK_USB_PID_STLINK_NUCLEO)
-                    || (desc.idProduct == STLINK_USB_PID_STLINK_32L_AUDIO)) {
-                sl->version.stlink_v = 2;
-            } else if ((desc.idProduct == STLINK_USB_PID_STLINK_V3_USBLOADER)
-                       || (desc.idProduct == STLINK_USB_PID_STLINK_V3E_PID)
-                       || (desc.idProduct == STLINK_USB_PID_STLINK_V3S_PID)
-                       || (desc.idProduct == STLINK_USB_PID_STLINK_V3_2VCP_PID)) {
-                sl->version.stlink_v = 3;
-            }
+        if ((serial == NULL) || (*serial == 0))
+            break;
 
-            if ((serial == NULL) || (*serial == 0))
-                break;
+        if (sl->serial_size < 0)
+            continue;
 
-            if (sl->serial_size < 0)
-                continue;
+        if (memcmp(serial, &sl->serial, sl->serial_size) == 0)
+            break;
 
-            if (memcmp(serial, &sl->serial, sl->serial_size) == 0)
-                break;
+		libusb_close(handle);
```

Comment:
@chenguokai: I think I found the reason - it's a bug:` libusb_close(handle);` is called at the wrong point in the routine. The order has changed here.

Reply:
In 09ea99aea51f741ce9c8c0e2ad755d40b3d86021 it was called _before_

`/* could not open device, continue */ if (ret) continue;`

while it is called much later in the version above. It does make a change. In a first try everything seemed to work as expected when placing this call before

`if ((serial == NULL) || (*serial == 0)) break;`,

but I'm not sure yet, if this _really_ is the correct place where it should go. We need to investigate carefully in order not to break anything else...

Reply:
Taking a closer look reveals that the command order in the function `stlink_t *stlink_open_usb(...)` is mixed up between:

`sl->serial_size = libusb_get_string_descriptor_ascii(handle, desc.iSerialNumber, (unsigned char *)sl->serial, sizeof(sl->serial));`

and

`WLOG ("Couldn't find %s ST-Link devices\n", (devBus && devAddr) ? "matched":"any");`

holding duplicated lines presumably deriving from old and new versions. This section clearly needs to be cleaned up.

This is one of these typical problems that are likely to appear, if branches are not updated regularly with changes from the main development branch. One ends up with a few merges which introduce loads of changes at once and in the end troubles arise in figuring out what went wrong. Clearly not a good practise. :disappointed:

Reply:
This should now be fixed with def770b0c5ea841a69e7828a11baa4c306d2ce6a.